src/util/submit-addon.js

```diff
@@ -0,0 +1,336 @@
+
+    log.info(`Fetching URL: ${url}`);
+    let headers = {
+      Authorization: `JWT ${authToken}`,
+      Accept: 'application/json',
+    };
+    if (typeof body === 'string') {
+      headers = {
+        ...headers,
+        'Content-Type': 'application/json',
+      };
+    }
+    return fetch(url, { method, body, headers });
+  }
+
+  async downloadSignedFile(
+    fileUrl: string,
+    addonId: string
+  ): Promise<Result> {
+    const filename = fileUrl.split('/').pop(); // get the name from fileUrl
+    const dest = `${this.downloadDir}/${filename}`;
+    const response = await this.fetch(fileUrl);
+    if (!response.ok || !response.body) {
+      log.info(`Download of signed xpi failed: ${response.status}.`);
+      throw new Error(`Downloading ${filename} failed`);
+    }
+    await this.saveToFile(response.body, dest);
+    return {
+      id: addonId,
+      downloadedFiles: [filename],
+    };
+  }
+
+  async saveToFile(
+    contents: typeof Response.body, destPath: string): Promise<any> {
+    return promisify(pipeline)(contents, createWriteStream(destPath));
+  }
+
+  async postNewAddon(
+    xpiPath: string,
+    channel: string,
+    metaDataJson: Object
+  ): Promise<Result> {
+    const uploadUuid = await this.doUploadSubmit(xpiPath, channel);
+
+    const versionObject =
+      channel === 'listed' ? 'current_version' : 'latest_unlisted_version';
+    const { guid: addonId, [versionObject]: {id: newVersionId}} =
+      await this.doNewAddonSubmit(uploadUuid, metaDataJson);
+
+    const fileUrl = await this.waitForApproval(addonId, newVersionId);
+
+    return this.downloadSignedFile(fileUrl, addonId);
+  }
+
+  async putVersion(
+    xpiPath: string,
+    channel: string,
+    addonId: string,
+    metaDataJson: Object
+  ): Promise<Result> {
+    const upload_uuid = await this.doUploadSubmit(xpiPath, channel);
+
+    await this.doNewAddonOrVersionSubmit(addonId, upload_uuid, metaDataJson);
+
+    const url =
+      `${this.apiUrl}addon/${addonId}/versions/?filter=all_with_unlisted`;
+    const {results: [{id: newVersionId}]} =
+      await this.fetchJson(url);
+
+    const fileUrl = await this.waitForApproval(addonId, newVersionId);
+
+    return this.downloadSignedFile(fileUrl, addonId);
+  }
+}
+
+type signAddonParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiHost: string,
+  // apiProxy,
+  timeout: number,
+  id?: string,
+  xpiPath: string,
+  downloadDir: string,
+  channel: string,
+|}
+
+export async function signAddon({
+  apiKey,
+  apiSecret,
+  apiHost,
+  // apiProxy,
+  timeout,
+  id,
+  xpiPath,
+  downloadDir,
+  channel,
+}: signAddonParams): Promise<Result> {
+  try {
+    const stats = await fsPromises.stat(xpiPath);
+
+    if (!stats.isFile()) {
+      throw new Error(`not a file: ${xpiPath}`);
+    }
+  } catch (statError) {
+    throw new Error(`error with ${xpiPath}: ${statError}`);
+  }
+
+  const client = new Client({
+    apiKey,
+    apiSecret,
+    apiHost,
+    // apiProxy,
```

Comment:
```suggestion
```

I don't think this is terribly useful. If you want, you can add a comment with a link to the issue to add support for a proxy in this client (not here but somewhere, maybe in the constructor).